common/src/main/java/io/netty/util/internal/Murmur3.java

```diff
@@ -0,0 +1,464 @@
+                }
+                @Override
+                public long byteArrayBaseOffset() {
+                    return BYTE_ARRAY_BASE_OFFSET;
+                }
+
+                // The logic below is pretty low level but has been shown to be beneficial in benchmarking.
+                // It essentially avoids an extra shift per byte for LONG and casting to/from long/byte/int.
+                // The same optimization was tried for the INT case but it was slightly worse for smaller array
+                // sizes (~100) and did not improve until bigger array sizes (>1000).
+                private static final long LONG_8_MASK = 0xFF0000;
+                private static final long LONG_8_A = ((long) 'A') << 16;
+                private static final long LONG_8_Z = ((long) 'Z') << 16;
+                private static final long LONG_8_LOWER_CASE = (((long) 'a') << 16) - LONG_8_A;
+                private static final long LONG_16_MASK = 0xFF00000000L;
+                private static final long LONG_16_A = ((long) 'A') << 32;
+                private static final long LONG_16_Z = ((long) 'Z') << 32;
+                private static final long LONG_16_LOWER_CASE = (((long) 'a') << 32) - LONG_16_A;
+                private static final long LONG_24_MASK = 0xFF000000000000L;
+                private static final long LONG_24_A = ((long) 'A') << 48;
+                private static final long LONG_24_Z = ((long) 'Z') << 48;
+                private static final long LONG_24_LOWER_CASE = (((long) 'a') << 48) - LONG_24_A;
+                private long offset8(long raw) {
+                    long masked = raw & LONG_8_MASK;
+                    return ((LONG_8_A <= masked && masked <= LONG_8_Z) ? masked + LONG_8_LOWER_CASE : masked) >>> 8;
+                }
+                private long offset16(long raw) {
+                    long masked = raw & LONG_16_MASK;
+                    return ((LONG_16_A <= masked && masked <= LONG_16_Z) ? masked + LONG_16_LOWER_CASE : masked) >>> 16;
+                }
+                private long offset24(long raw) {
+                    long masked = raw & LONG_24_MASK;
+                    return ((LONG_24_A <= masked && masked <= LONG_24_Z) ? masked + LONG_24_LOWER_CASE : masked) >>> 24;
+                }
+            }, DEFAULT_SEED);
+        } else {
+            logger.info("Unsafe not found. Murmur3 Unsafe char[] lookups will be slower.");
+        }
+        STRING_VALUE_FIELD_OFFSET = stringValueFieldOffset;
+        STRING_BUILDER_VALUE_FIELD_OFFSET = stringBuilderValueFieldOffset;
+        INSTANCE_UNSAFE = unsafe;
+        INSTANCE_UNSAFE_CASE_INSENSITIVE = unsafeAsciiLowerCase;
+
+        INSTANCE_SAFE = new Murmur3(new AbstractNativeTypeAccessor() {
+            @Override
+            public int getInt(byte[] data, long i) {
+                final int intI = (int) i;
+                return data[intI + 3] << 24 |
+                        (data[intI + 2] & 0xFF) << 16 |
```

Comment:
Refresh my memory ... why do we need the `0xFF` mask? Does the `<<` operation behave the differently without it for signed bytes?